### DISCUSSION TITLE
add parameter to onTransition to prevent calling listener immediately

### DISCUSSION THREAD

**Discussion by cmbro** [statelyai/xstate#1707] [Ideas]
I had an issue that took me bit to figure out due to not realizing that onTransition will immediately call the listener.  To work around, I am saving state outside my listener (and updating everytime the transition listener is called) and comparing the state sent to the listener with the "cached" state:

`if (state === session.state) {
  return;
}`

It would be useful to be able to pass false to onTransition to prevent that initial call:

`service.onTransition(onStateChange, false)`

By default, it would work the same as now, but the extra parameter could specify whether the listener is called immediately.  Thoughts?

**Comment by Andarist** [comment]
Just my 2 cents: I was initially surprised as well that this callback is being called upon registration. Usually, I expect `on*` handlers to be called upon actual change.

**Reply by davidkpiano** [reply]
My original intention was for this to behave like a [`ReplaySubject` in RxJS](https://rxjs-dev.firebaseapp.com/api/index/class/ReplaySubject), because it is important to know what the current state is without having to rely on implementation details for getting the latest state value.

**Reply by cmbro** [reply]
I think most of the time, it is best the way it is and you are correct, getting that current state in the listener is perfect.  My situation is on the server setting listeners to a running service in a http request handler.  I set the listener, send the event (from the client) to the interpreter service, and then render a view back to the client in the onTransition listener.  Client side as an SPA, it was perfect.  But after switching to server side rendering, I need to filter out that initial call when the listener is set in the request handler.  My work around does handle this, but I was thinking it wouldn't be necessary if that was optional.  Eventually, I may serialize state and restart the service every request in which case this won't matter since I can set the handler before starting the service.  (I only have a 2 to 3 dozen sessions at a time to worry about, so I am not needing to do that at this point... and I am curious to see what changes V5 may bring in regards to serialization)
